## 03-Lab  - Regularization- Ridge, Lasso and ElasticNet

In this lab, you will learn how to implement Regularization techniques using sklearn and Statsmodels.  You will work with Credit dataset - description provided below .
The dataset contains information about credit information of few people and their other details. The goal is to predict the Bank Balance of each individual based on their credit information. 
Your task is divided into three parts: 
1) Part 1 : Read the dataset, convert categorical dataset to dummy indicies so we can use them for regression and Visual the dataset.<br>
2) Perform a simple polynomaial regression using Sklearn and compare it with Ridge, Lasso and ElasticNet regression.  <br>
3) Use Statsmodels OLS regression to perform linear regression and regression using Elastic Net <br>

Data is availabe in : https://raw.githubusercontent.com/colaberry/DSin100days/master/data/Credit.csv

In [1]:
# Import all libraries
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression, Ridge, Lasso, ElasticNet
from sklearn.metrics import mean_squared_error
from sklearn.preprocessing import PolynomialFeatures, StandardScaler
import numpy as np 
import matplotlib.pyplot as plt
import seaborn as sns
import statsmodels.api as sm

#Write your code here or in other code cells down

## Part 1: Data Cleaning and Prep
In this part you will be importing the data. Reading the data into a dataframe. You will need to use pandas to create dummy indicies covert categorical variable dummy columns. You will then generate the test train splits for the dataset. 


In [2]:
# Import the dataset. We have provided the file location in file_loc. Use pandas to read the csv file
# make sure you include na_values as '?'. Make sure you also do dropna to remove nan values. 
# You will have to show by df.info() and df.head() (in that order) to get the output shown.
file_loc = 'https://raw.githubusercontent.com/colaberry/DSin100days/master/data/Credit.csv'
credit = "####"
credit"####"
credit"####"


<class 'pandas.core.frame.DataFrame'>
Int64Index: 400 entries, 0 to 399
Data columns (total 12 columns):
Unnamed: 0    400 non-null int64
Income        400 non-null float64
Limit         400 non-null int64
Rating        400 non-null int64
Cards         400 non-null int64
Age           400 non-null int64
Education     400 non-null int64
Gender        400 non-null object
Student       400 non-null object
Married       400 non-null object
Ethnicity     400 non-null object
Balance       400 non-null int64
dtypes: float64(1), int64(7), object(4)
memory usage: 40.6+ KB


Unnamed: 0   Income  Limit  Rating  Cards  Age  Education  Gender Student  \
0           1   14.891   3606     283      2   34         11    Male      No   
1           2  106.025   6645     483      3   82         15  Female     Yes   
2           3  104.593   7075     514      4   71         11    Male      No   
3           4  148.924   9504     681      3   36         11  Female      No   
4           5   55.882   4897     357      2   68         16    Male      No   

  Married  Ethnicity  Balance  
0     Yes  Caucasian      333  
1     Yes      Asian      903  
2      No      Asian      580  
3      No      Asian      964  
4     Yes  Caucasian      331

<img src="../../../images/regularization_1.png">

In [3]:
# You will have to convert categorical variables to dummy indicies. Pandas lets us do this very easily 
# with a function called get_dummies. Use that to get dummy indicies. Get the dtypes of the df with 
# dummy variables and get the values of those dtypes. This what we want you to print as a result.
credit_with_dummies = "####"
check_all_variables ="####"
print("Data type of each columns is: \n{}".format(check_all_variables))

# You need to drop unnamed column from the dataframe. Look up the columns and ensure it this column is dropped
# Note: If you do not drop the column then your results will not match from this point on.
credit_with_dummies = "####"

Data type of each columns is: 
[dtype('int64') dtype('float64') dtype('int64') dtype('int64')
 dtype('int64') dtype('int64') dtype('int64') dtype('int64')
 dtype('uint8') dtype('uint8') dtype('uint8') dtype('uint8')
 dtype('uint8') dtype('uint8') dtype('uint8') dtype('uint8')
 dtype('uint8')]


Data type of each columns is: <br>
[dtype('int64') dtype('float64') dtype('int64') dtype('int64') <br>
 dtype('int64') dtype('int64') dtype('int64') dtype('int64') <br>
 dtype('uint8') dtype('uint8') dtype('uint8') dtype('uint8') <br>
 dtype('uint8') dtype('uint8') dtype('uint8') dtype('uint8') <br>
 dtype('uint8')] <br>

In [4]:

# checking for missing values 

# We need to check if nay of the columns have any missing values. You can do that by using isnull() and any() 
# functions. Both are methods that can be called on a df. For example you can do  df.isnull(). 
check_missing = "####"
print("Check if any values are null we get: \n{}".format(check_missing))

Check if any values are null we get: 
[False False False False False False False False False False False False
 False False False False]


Check if any values are null we get: <br>
[False False False False False False False False False False False False  <br>
 False False False False False] <br>

Now that we have imported the data our first step is prepare the data for Linear regression. For this we need to split our data into the features and target. Below you will do that. 

In [5]:
# You will need to drop the traget column. In the question description we 
# have mentioned the variable you will be regressing for, so drop that column 
# to get all the features in the dataset.
X ="####"

# The column we dropped above will be the target hence y must equal that column.
y ="####"



In [6]:
# Use seaborn's pairplot to show how each of the features vary with each other
# note this operation may take upto a minute since we have a lot of columns
# Make sure you pairplot X and not y since y will only contain 1 column 
fig = "####"
plt.show()

<img src="../../../images/regularization_2.png">

From the pair plot we can see that we have a strong correlation between ```ratings``` and ```Limit``` variables, hence we will drop one of these. We will next drop the ```Limit``` variable. 

In [7]:
# Drop Limit variable from X
X ="####"


In [8]:
# Next use standard scalar to scale and center the dataset.  
scaler = "####"

# Create a new dataset where the data is the fit tranformed output of X and 
# columns are the X.columns.
scaled_X ="####"
scaled_X.head()

Income    Rating     Cards       Age  Education  Gender_ Male  \
0 -0.861583 -0.465539 -0.699130 -1.257674  -0.784930      1.035635   
1  1.727437  0.828703  0.031032  1.528451   0.496588     -0.965592   
2  1.686756  1.029311  0.761194  0.889964  -0.784930      1.035635   
3  2.946152  2.110003  0.031032 -1.141586  -0.784930     -0.965592   
4  0.302928  0.013331 -0.699130  0.715831   0.816968      1.035635   

   Gender_Female  Student_No  Student_Yes  Married_No  Married_Yes  \
0      -1.035635    0.333333    -0.333333   -0.795395     0.795395   
1       0.965592   -3.000000     3.000000   -0.795395     0.795395   
2      -1.035635    0.333333    -0.333333    1.257237    -1.257237   
3       0.965592    0.333333    -0.333333    1.257237    -1.257237   
4      -1.035635    0.333333    -0.333333   -0.795395     0.795395   

   Ethnicity_African American  Ethnicity_Asian  Ethnicity_Caucasian  
0                   -0.573501        -0.585049             1.005013  
1                   -0.573501         1.709260            -0.995012  
2                   -0.573501         1.709260            -0.995012  
3                   -0.573501         1.709260            -0.995012  
4                   -0.573501        -0.585049             1.005013

<img src="../../../images/regularization_3.png">

In [9]:
# Use train test split to split the dataset into training set and testing set. 
# use a test size of 0.3 and random_state of 2
X_train, X_test, y_train, y_test= "####"

# Next declare an instance of Polynomial features class from sklearn with degree 2. We will
# use this to create polynomial features
pf = "####"

# you will need to fit transform the training data features and the testing set features to generate polynomial feature.
poly_train_X = "####"
poly_test_X = "####"


# Add a new axis to y_train and y_test so you can perform linear regression 
y_train = "####"
y_test = "####"


# Shape of the polynomial traning and testing sets
print("Shape of transformed training set: {}".format(poly_train_X.shape))
print("Shape of transformed testing set: {}".format(poly_test_X.shape))

Shape of transformed training set: (280, 120)
Shape of transformed testing set: (120, 120)


Shape of transformed training set: (280, 136) <br>
Shape of transformed testing set: (120, 136)

## Part 2: Linear, Ridge, Lasso and ElasticNet regression using Sklearn 

Next we are going to use the transformed data to run 

In [30]:
# Delcare an instance of Linear Regression class from Sklearn, make sure you set fit_interpet to False
# We set the intercept to false since we have already centered the data using standard scalar.
lr = "####"

# Fit the training data with polynomial features to the linear regression model. 
lr."####"

# Predict on the test set. Remember the test set must also be transformed to polynomial features. 
y_pred = "####"

# Calculate the mean squared error using the function from Sklearn 
mse = "####"
print("Mean squared error: {:.2f}".format(mse))

Mean squared error: 7149.13


Mean squared error: 7149.13


Next we are going to try Ridge and Lasso regression from sklearn 

In [24]:
# Delcare an instance of Ridge class, set alpha =1.2 normalize=False and max_iter=10000.
lr_ridge = "####"

# similar to linear regression fit the polynomial features training set. 
"####"

# Run pedict on the test set.
"####"
# Calculate the mean squared error using the function from Sklearn.
mse = "####"
print("Mean squared error: {:.2f}".format(mse))

Mean squared error: 7381.96


Mean squared error: 7381.96


In [23]:
# Delcare an instance of Lasso class, set alpha =1.1 normalize=False and max_iter=10000.
lr_lasso = "####"

# similar to linear regression fit the polynomial features training set. 
lr_lasso."####"

# Run pedict on the test set.
y_pred = "####"

# Calculate the mean squared error using the function from Sklearn. 
mse = "####"
print("Mean squared error: {:.2f}".format(mse))

Mean squared error: 6673.42


Mean squared error: 6673.42


In [25]:
# Delcare an instance of  ElasticNet , set alpha =1.5, l1_ratio=0.5,  fit_intercept=False and random_state=0.
lr_en = "####"
lr_en.fit(poly_train_X, y_train)

# Predictions work the same way as sklearn. Run predict on the test set. 
pred_y = lr_en.predict(poly_test_X)

# Calculate the mean squared error using the function from Sklearn. 
mse ="####"
print("Mean squared error: {:.2f}".format(mse))

Mean squared error: 89630.78


Mean squared error: 89630.78


You will find that we performed better with Lasso regression than with Ridge. This is because Lasso typically imposes heavy penalty on points which deviate from away from the mean. 
Next part we look at how we can run linear regression and elastic net in statsmodels. 

## Part 3: LR and elastic net using Statsmodels
Next we are going to use statsmodels OLS regression perform linear regression. You may want to check the documentation for statsmodels [here](https://www.statsmodels.org/stable/regression.html)


In [14]:
# Using sm.OLS fit a regression model. The sytax for statsmodels is a bit different
# you will need chain fit() to sm.OLS(y,X). Remember you X must be composed of polynomial features.
st_model = "####"

# Predictions work the same way as sklearn. Run predict on the test set. 
pred_y = "####"

# Calculate the mean squared error using the function from Sklearn.
mse = "####"
print("Mean squared error: {:.2f}".format(mse))

Mean squared error: 7143.16


Mean squared error: 7143.16


Next we will train a statsmodels model using elastic net. To do this will be using fit_regualized option from statsmodel. This option is a method of the OLS model that we train for linear regression i.e we just replace the fit() method in OLS with fit regularized. You can read the documentation [here](https://www.statsmodels.org/stable/generated/statsmodels.regression.linear_model.OLS.fit_regularized.html) to better understand what each of the arguments do. 

In [21]:
# Fit a elastic net model to the training data. fit_regularized is a method of sm.OLS(y,X). Remember to set
# the following arguments for fit regularized: method='elastic_net', alpha=1.5 and L1_wt as 0.5 
rmodel = "####"

# Predictions work the same way as sklearn. Run predict on the test set 
pred_y = "####"

# Calculate the mean squared error using the function from Sklearn 
mse = "####"
print("Mean squared error: {:.2f}".format(mse))

Mean squared error: 90526.54


Mean squared error: 90526.54


You will find that the ElasticNet results for Sklearn and Statsmodels will be pretty much be the same. The same will apply for Linear Regression, there will be some differences between the results due to the way the calucations are done in the backend by results should not be off by an order of magnitude.

You might be wondering why did we even both with Statsmodels? Well run the command in the cell below and you should see various statistical measures and values for the linear regression model. One of the most important quantities you will get from this is the p-value of the fit(```Prob (F-statistic):```) This p-value is derived from an F-test which tells you how significant your fit is. If the fit is less than 0.05 you can state that you fit is significant i.e the test states that changes in your traget variable depend on the changes in your feature, if the p-value is greater than 0.05 then we conclude that the test cannot determine if the target varaible changes due to changes in the features. 

You will also notice that we have a lot of features, this is because we are using the polynomial features and not the dataset features. The summary will also give us access to the 25% and 95% confidence intervals each of the coefficients.  


In [27]:
st_model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.981
Model:                            OLS   Adj. R-squared:                  0.976
Method:                 Least Squares   F-statistic:                     190.6
Date:                Tue, 05 Jan 2021   Prob (F-statistic):          2.48e-160
Time:                        14:08:27   Log-Likelihood:                -1560.8
No. Observations:                 280   AIC:                             3242.
Df Residuals:                     220   BIC:                             3460.
Df Model:                          59                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         41.7204      1.264     32.995      0.000      39.228      44.212
x1          -282.6646      9.922    -28.489      0.000    -302.218    -263.111
x2           618.3454      7.755     79.736      0.000     603.062     633.629
x3            -2.0472      5.751     -0.356      0.722     -13.381       9.287
x4           -16.8637      4.903     -3.440      0.001     -26.526      -7.201
x5            -3.9881      5.190     -0.768      0.443     -14.216       6.240
x6            -3.2633      2.427     -1.345      0.180      -8.046       1.519
x7             3.2633      2.427      1.345      0.180      -1.519       8.046
x8             8.7963      0.535     16.437      0.000       7.742       9.851
x9            -8.7963      0.535    -16.437      0.000      -9.851      -7.742
x10          -17.4252      2.143     -8.130      0.000     -21.649     -13.201
x11           17.4252      2.143      8.130      0.000      13.201      21.649
x12          -13.3873      1.381     -9.692      0.000     -16.110     -10.665
x13          -11.1752      1.430     -7.817      0.000     -13.992      -8.358
x14           21.2968      1.848     11.526      0.000      17.655      24.938
x15           52.0410     12.418      4.191      0.000      27.568      76.514
x16         -178.1063     23.913     -7.448      0.000    -225.234    -130.979
x17            4.8286      9.372      0.515      0.607     -13.641      23.298
x18            6.8128      8.843      0.770      0.442     -10.616      24.241
x19           -5.8480      9.133     -0.640      0.523     -23.847      12.151
x20           -0.2192      4.205     -0.052      0.958      -8.507       8.068
x21            0.2192      4.205      0.052      0.958      -8.068       8.507
x22           20.9759      4.507      4.654      0.000      12.093      29.858
x23          -20.9759      4.507     -4.654      0.000     -29.858     -12.093
x24            1.4720      4.478      0.329      0.743      -7.352      10.296
x25           -1.4720      4.478     -0.329      0.743     -10.296       7.352
x26           -5.7664      6.071     -0.950      0.343     -17.731       6.199
x27           10.7206      6.739      1.591      0.113      -2.561      24.002
x28           -4.3684      5.131     -0.851      0.396     -14.481       5.744
x29          157.9286     13.613     11.601      0.000     131.100     184.757
x30          -15.1313      8.982     -1.685      0.093     -32.833       2.571
x31          -12.6743      7.840     -1.617      0.107     -28.125       2.776
x32            9.6102      8.140      1.181      0.239      -6.432      25.653
x33            0.0451      3.913      0.012      0.991      -7.666       7.756
x34           -0.0451      3.913     -0.012      0.991      -7.756       7.666
x35          -28.9701      4.245     -6.824      0.000     -37.336     -20.604
x3